url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/math/statistics/Generic_Statistics.java

```diff
@@ -0,0 +1,850 @@
+        BigDecimal moment3 = BigDecimal.ZERO;
+        BigDecimal moment4 = BigDecimal.ZERO;
+
+        // Deal with special cases
+        int n = data.size();
+        BigDecimal n_BigDecimal = BigDecimal.valueOf(n);
+        if (n < 2) {
+            secondOrderStatistics[0] = BigDecimal.ZERO;
+            secondOrderStatistics[1] = BigDecimal.ZERO;
+            secondOrderStatistics[2] = BigDecimal.ZERO;
+            secondOrderStatistics[3] = BigDecimal.ZERO;
+            secondOrderStatistics[4] = BigDecimal.ZERO;
+        }
+        BigDecimal value;
+        BigDecimal meandiff;
+        Iterator<BigDecimal> ite = data.iterator();
+        while (ite.hasNext()) {
+            value = ite.next();
+            meandiff = value.subtract(summaryStatistics_0[1]);
+            moment1 = moment1.add(meandiff);
+            moment2 = moment2.add(meandiff.pow(2));
+            moment3 = moment3.add(meandiff.pow(3));
+            moment4 = moment4.add(meandiff.pow(4));
+        }
+        BigDecimal variance = Generic_BigDecimal.divideRoundIfNecessary(
+                moment2,
+                n_BigDecimal,
+                decimalPlaces,
+                roundingMode);
+        secondOrderStatistics[0] = moment1;
+        secondOrderStatistics[1] = moment2;
+        secondOrderStatistics[2] = moment3;
+        secondOrderStatistics[3] = moment4;
+        secondOrderStatistics[4] = variance;
+        try {
+        secondOrderStatistics[5] = Generic_BigDecimal.power(
+                variance,
+                BigDecimal.valueOf(0.5d),
+                decimalPlaces,
+                roundingMode);
+        } catch (UnsupportedOperationException e) {
+            // A terrible hack!
+            secondOrderStatistics[5]  = variance.divide(BigDecimal.valueOf(2));
+        }
+        result[1] = secondOrderStatistics;
+//        System.out.println("Moment 1 " + moment1.toPlainString());
+//        System.out.println("Moment 2 " + moment2.toPlainString());
+//        System.out.println("Moment 3 " + moment3.toPlainString());
+//        System.out.println("Moment 4 " + moment4.toPlainString());
+//        System.out.println("Variance " + variance.toPlainString());
+//        System.out.println("Standard Devation /n " + secondOrderStatistics[5].toPlainString());
+        return result;
+    }
+
+    /**
+     * There is no universal agreement on calculating quartiles
+     * http://en.wikipedia.org/wiki/Quartile
+     * @param data
+     * @param decimalPlaces
+     * @param roundingMode
+     * @return 
+     * result[0] = sum;
+     * result[1] = mean;
+     * result[2] = median;
+     * result[3] = q1;
+     * result[4] = q3;
+     * result[5] = mode;
+     * result[6] = min;
+     * result[7] = max;
+     * result[8] = numberOfDifferentValues;
+     * result[9] = numberOfDifferentValuesInMode;
+     * result[10] = numberOfSameValuesInAnyPartOfMode;
+     */
+    public static BigDecimal[] getSummaryStatistics_0(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 55. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 181-209, lines 210-238, lines 239-279 into separate methods.